trieste/acquisition/function.py

```diff
@@ -509,6 +512,156 @@ def prepare_acquisition_function(
+        the original paper, but is applicable to different non-dominated partition algorithm
+    2. The Psi and nu function in the original paper is defined for a maximization problem, to
+       make use of the same notation for easier following, we inverse our problem (as maximization)
+       to make use of the same equation
+
+    :param model: The model of the objective function.
+    :param pareto: Pareto class
+    :param reference_point The reference point for calculating hypervolume
+    :return The expected_hv_improvement acquisition function modified for objective
+        minimisation. This function will raise :exc:`ValueError` or
+        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+    """
+
+    def acquisition(x: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(x, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        normal = tfp.distributions.Normal(
+            loc=tf.zeros(shape=1, dtype=x.dtype), scale=tf.ones(shape=1, dtype=x.dtype)
+        )
+
+        def Psi(a, b, mean, std) -> TensorType:
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return std * normal.prob((b - mean) / std) + (mean - a) * (
+                1 - normal.cdf((b - mean) / std)
+            )
+
+        def nu(lb, ub, mean, std) -> TensorType:
+            """
+            A special expected improvement given fixed improvement,
+            defined at Eq. 25 in :cite:`yang2019efficient`
+            :param lb: shape [..., num_cells, out_dim] slice lower bound
+            :param ub: shape [..., num_cells, out_dim] slice upper bound
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return (ub - lb) * (1 - normal.cdf((ub - mean) / std))
+
+        candidate_mean, candidate_var = model.predict(tf.squeeze(x, -2))
+        candidate_std = tf.sqrt(candidate_var)
+
+        # calc ehvi assuming maximization
+        neg_candidate_mean = -tf.expand_dims(candidate_mean, 1)  # [..., 1, out_dim]
+        candidate_std = tf.expand_dims(candidate_std, 1)  # [..., 1, out_dim]
+
+        lb_points, ub_points = pareto.hypercell_bounds(
+            tf.constant([-inf] * candidate_mean.shape[-1], dtype=x.dtype), reference_point
+        )
+
+        neg_lb_points, neg_ub_points = -ub_points, -lb_points
+
+        neg_ub_points = tf.minimum(
+            neg_ub_points, 1e10
```

Comment:
The truncation is more for numerical stability, as the the gradient of CDF of `inf` will cause `nan` issue IIRC. So a naive fix is applied here like truncating the upper bound to avoid the direct operation on `inf`. Please suggest on this point if any better way could be of use.